src/main/java/com/conveyal/taui/controllers/AggregationAreaController.java

```diff
@@ -99,45 +112,69 @@ public static AggregationArea createAggregationArea (Request req, Response res)
 
         ShapefileReader reader = new ShapefileReader(shpFile);
 
-        List<Geometry> geometries = reader.stream().map(f -> (Geometry) f.getDefaultGeometry()).collect(Collectors.toList());
-        UnaryUnionOp union = new UnaryUnionOp(geometries);
-        Geometry merged = union.union();
+        List<SimpleFeature> features = reader.wgs84Stream().collect(Collectors.toList());
+
+        Map<String, Geometry> areas = new HashMap<>();
+
+        if (Boolean.parseBoolean(req.params("union"))) {
+            List<Geometry> geometries = features.stream().map(f -> (Geometry) f.getDefaultGeometry()).collect(Collectors.toList());
+            UnaryUnionOp union = new UnaryUnionOp(geometries);
+            // Use provided name directly for this single unioned feature
+            areas.put(maskName, union.union());
+        } else {
+            // Use provided name to read a name for each feature
+            try {
+                features.stream().forEach(f -> areas.put(f.getProperty(maskName).getValue().toString(), (Geometry) f.getDefaultGeometry()));
+            } catch (NullPointerException e) {
```

Comment:
Other things could cause a NullPointerException. We should probably split the operation of reading the property out into another line, and check specifically for a missing value rather than a general exception.